Support access of some tile constants with Cesium3DTileStyle

SunBlack opened the pull request:
Support access of RTC_CENTER, QUANTIZED_VOLUME_SCALE and QUANTIZED_VOLUME_OFFSET by Cesium3DTileStyle.

Example:
```
new Cesium3DTileStyle({
	"color": "rgb((${RTC_CENTER}.z + ${POSITION}.z) % 255, 0, 0)"
});
```

Related issue: #4654

lilleyse commented:
I wonder if this could go a step further have a generic `CENTER` property that essentially equals the bounding sphere's center. That way you wouldn't need separate styles depending on if your tiles are rtc, quantized, or use a tile transform.

`QUANTIZED_VOLUME_SCALE` feels too specific to me, and wouldn't be needed anymore to derive the center. Is there a good use case beyond that?

SunBlack commented:
I tried to colorize a point cloud with a height scale. Without access to this variables it is not possible to apply a global gradient and not a gradient per tile. This didn't succeeded, because the value z is not the height above the earth because of the Cartesian coordinate system.

But you are right, maybe it is better to implement a more generic way, so you don't need to know how data are stored. Nevertheless: A center variable won't work, because of different shift of position. Maybe it is an idea that ${POSITION} isn't anymore the raw position, but the absolute position. So instead of `POSITION := position` in case an RTC_Center exists ´POSITION := RTC_CENTER + position` and if not `POSITION := POSITION_QUANTIZED * QUANTIZED_VOLUME_SCALE + QUANTIZED_VOLUME_OFFSET`? I'm not even know which values ${POSITION} currently contains in case positons are quantized. But even with this change I don't know if ${POSITION} is a really helpful variable because of Cartesian coordinate system (position in GPS format would be more helpful, so you have a usable z coordinate).

SunBlack commented:
I have adjusted available semantic values so the user don't need to think about how position is stored in pnts. Additionally I added DISTANCE_TO_EARTHCENTER - this method is really simple and could be also implement by normal style syntax (okay, length is currently not implemented) - but if you have an idea how I can implement DISTANCE_TO_ELLIPSOID it would make sense to have this value.

lilleyse commented:
Instead of passing in `rtcCenter` and `quantizedVolumeOffset`, could you pass in the draw command's `modelMatrix` and then multiply against the point? This will ensure it works for transformed tiles (rotation and scale included). Overall though the approach looks good to me for `POSITION_ABSOLUTE`.

I'm a little less sure about `DISTANCE_TO_EARTHCENTER` and `DISTANCE_TO_ELLIPSOID`. Tilesets or even just other 3D Tiles implementations may not always exist in a geospatial context, so these seem too specific to make part of the spec. Maybe once the spec has the concept of extensions these would make sense to include. 

So for now computing with `length` would be the way to go. `length` and other vector operations are coming pretty soon (https://github.com/AnalyticalGraphicsInc/cesium/issues/4865).

`DISTANCE_TO_ELLIPSOID` is definitely harder. You can check out the `Cartographic.fromCartesian` implementation or other distance-from-ellipsoid code. We don't have this in GLSL yet, and it may not be possible to inline in a style. I feel like storing heights as a batch table property is best for now, though with a bit of a data overhead.

SunBlack commented:
Good idea to use the draw command's `modelMatrix`. Now the code is really simple. Are the matrices already accessible via Cesium3DTileStyle? If yes `POSITION_ABSOLUTE` is maybe not necessary.

I removed `DISTANCE_TO_EARTHCENTER` as long as I don't have a working implementation of `DISTANCE_TO_ELLIPSOID` - I hope the implementation of `length` will coming soon.

lilleyse commented:
The matrix isn't accessible by the style, so your approach seems good to me.

lilleyse commented:
Can you also open a PR to add this to https://github.com/AnalyticalGraphicsInc/3d-tiles/tree/master/Styling#point-cloud?

Personally I prefer `POSITION_ABSOLUTE` over `ABSOLUTE_POSITION`.

lilleyse commented:
One more thing, could you send over a CLA so we can review/merge these changes? https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#contributor-license-agreement-cla

SunBlack commented:
done

lilleyse commented:
It looks like my suggestion to rename to `POSITION_ABSOLUTE` exposed an issue where `getStyleableSemantics` tried to match it to just `POSITION`.

To fix this, `getStyleableSemantics` can be removed and incorporated into `getStyleableProperties`, which could take three arguments, `source`, `properties`, and `semantics`.

SunBlack commented:
A parameter named `semantics` isn't good, because of global variable with same name. So I changed code a bit. `getStyleableProperties` now collects all properties, independent of its type (default or user defined property). Later I filter this list to have only `userProperties`. Furthermore I extended Regex  `modifyStyleFunction`, to fix the problem of `POSITION` vs `POSITION_ABSOLUTE`. The user can now choose freely property name, as long they are not default properties (in previous implementation the property name could not start with a name of a default property).

lilleyse commented on Source/Scene/PointCloud3DTileContent.js:
```diff
@@ -807,33 +807,21 @@ define([
     function getStyleableProperties(source, properties) {
         // Get all the properties used by this style
-        var regex = /czm_tiles3d_style_(\w+)/g;
+        var regex = /czm_tiles3d_style_([\w_]+)/g;
```
`\w` includes `_`

lilleyse replied:
Same comment below.

SunBlack replied:
You are right - had forgot that ´\w` includes `_`

lilleyse commented on Source/Scene/PointCloud3DTileContent.js:
```diff
@@ -845,17 +833,17 @@ define([
+            var property = defaultProperties[i];
+            var styleName = 'czm_tiles3d_style_' + property;
+            var replaceName = property.toLowerCase();
+            source = source.replace(new RegExp(styleName + "([^\w_$])", 'g'), replaceName + "$1");
```
Style: use single quotes for strings.

lilleyse replied:
Can this regex be simplified? `(\w*)`?

SunBlack replied:
No, because after `styleName` no letter should follow. But I think I can remove `$` from it, because there should be always still a `;` after a property. And maybe I should write `\W` instead of `[^\w]`

Btw: with single quote I need to to write `\\W` instead of `\W`

lilleyse replied:
Ah right, got it now.

lilleyse commented on Source/Scene/PointCloud3DTileContent.js:
```diff
@@ -907,27 +895,26 @@ define([
+        var usesColorSemantic = styleableProperties.indexOf('COLOR') >= 0;
+        var usesNormalSemantic = styleableProperties.indexOf('NORMAL') >= 0;
+		
+		// Split default properties from user properties
```
Fix indentation.

lilleyse commented on Source/Scene/PointCloud3DTileContent.js:
```diff
@@ -907,27 +895,26 @@ define([
+        var usesNormalSemantic = styleableProperties.indexOf('NORMAL') >= 0;
+		
+		// Split default properties from user properties
+        var userProperties = styleableProperties.filter(function(property) { return !defaultProperties.includes(property); });
```
Use `indexOf` instead, `includes` is not supported in IE.

lilleyse commented:
Ok, just one last thing - add yourself to `CONTRIBUTORS.md`

lilleyse commented:
Thanks @SunBlack

pjcozzi commented:
Thanks again for the 3D Tiles contributions, @SunBlack!  Are you able to share a bit about your project?  We could list it with [the spec](https://github.com/AnalyticalGraphicsInc/3d-tiles#whos-using-3d-tiles) if you want.  Reply here or email me at pcozzi@agi.com
